src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -241,77 +251,100 @@ public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext conte
                 throw new ArgumentNullException(nameof(context));
             }
 
-            return GetVariantAsync(feature, context, useContext: true, cancellationToken);
+            EvaluationEvent evaluationEvent = await EvaluateFeature(feature, context, useContext: true, cancellationToken);
+
+            return evaluationEvent.Variant;
         }
 
-        private async ValueTask<bool> IsEnabledWithVariantsAsync<TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
+        private async Task<EvaluationEvent> EvaluateFeature<TContext>(string feature, TContext context, bool useContext, CancellationToken cancellationToken)
         {
-            bool isFeatureEnabled = false;
+            var evaluationEvent = new EvaluationEvent
+            {
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
 
-            FeatureDefinition featureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false);
+            if (evaluationEvent.FeatureDefinition != null)
+            {
+                //
+                // Determine IsEnabled
+                evaluationEvent.IsEnabled = await IsEnabledAsync(evaluationEvent.FeatureDefinition, context, useContext, cancellationToken).ConfigureAwait(false);
 
-            VariantDefinition variantDefinition = null;
+                //
+                // Determine Variant
+                VariantDefinition variantDefinition;
 
-            if (featureDefinition != null)
-            {
-                isFeatureEnabled = await IsEnabledAsync(featureDefinition, appContext, useAppContext, cancellationToken).ConfigureAwait(false);
+                if (evaluationEvent.FeatureDefinition.Allocation == null || (!evaluationEvent.FeatureDefinition.Variants?.Any() ?? false))
+                {
+                    variantDefinition = null;
 
-                if (featureDefinition.Variants != null && featureDefinition.Variants.Any() && featureDefinition.Allocation != null)
+                    evaluationEvent.AssignmentReason = AssignmentReason.None;
+                }
+                else
                 {
-                    if (!isFeatureEnabled)
+                    if (!evaluationEvent.IsEnabled)
                     {
-                        variantDefinition = featureDefinition.Variants.FirstOrDefault((variant) => variant.Name == featureDefinition.Allocation.DefaultWhenDisabled);
+                        variantDefinition = evaluationEvent.FeatureDefinition.Variants.FirstOrDefault((variant) => variant.Name == evaluationEvent.FeatureDefinition.Allocation.DefaultWhenDisabled);
+
+                        if (variantDefinition != null)
```

Comment:
And if `variantDefinition` is null, then AssignmentReason seems to be the default value. Which looks to be AssignmentReason.None. But on line 280, the assignment reason is explicitly set to none. I expect we should stick to one strategy, either explicitly set assignment reason to none, or let it be none due to being the default.

Reply:
Sounds good. I vote for explicity setting the assignment reason to none.